File: src/main/java/me/dyzjct/kura/module/modules/xddd/SmartOffHand.java
package me.dyzjct.kura.module.modules.xddd;

import me.dyzjct.kura.module.Category;
import me.dyzjct.kura.module.Module;
import me.dyzjct.kura.module.modules.combat.AutoTotem;
import me.dyzjct.kura.setting.BooleanSetting;
import me.dyzjct.kura.setting.DoubleSetting;
import me.dyzjct.kura.setting.ModeSetting;
import me.dyzjct.kura.setting.Setting;
import me.dyzjct.kura.utils.NTMiku.TimerUtils;
import me.dyzjct.kura.utils.entity.CrystalUtil;
import me.dyzjct.kura.utils.entity.HoleUtil;
import java.util.ArrayList;
import net.minecraft.entity.Entity;
import net.minecraft.entity.item.EntityEnderCrystal;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.init.Items;
import net.minecraft.inventory.ClickType;
import net.minecraft.inventory.EntityEquipmentSlot;
import net.minecraft.item.Item;
import net.minecraft.item.ItemBlock;
import net.minecraft.item.ItemExpBottle;
import net.minecraft.item.ItemFood;
import net.minecraft.item.ItemStack;
import net.minecraft.item.ItemSword;
import org.lwjgl.input.Mouse;

@Module.Info(name="SmartOffHand", category=Category.XDDD, description="StupidOffHand")
public class SmartOffHand
        extends Module {
    public static TimerUtils timerUtils = new TimerUtils();
    public int totems;
    public int count;
    ModeSetting<Mode> mode = this.msetting("Mode", Mode.Crystal);
    Setting<Integer> delay = this.isetting("Delay", 0, 0, 1000);
    BooleanSetting totem = this.bsetting("SwitchTotem", true);
    DoubleSetting sbHealth = this.dsetting("Health", 11.0, 0.0, 36.0);
    BooleanSetting autoSwitch = this.bsetting("SwitchGap", true);
    ModeSetting<SMode> switchMode = this.msetting("GapWhen", SMode.RClick).b(this.autoSwitch);
    BooleanSetting elytra = this.bsetting("CheckElytra", true);
    BooleanSetting holeCheck = this.bsetting("CheckHole", false);
    DoubleSetting holeSwitch = this.dsetting("HoleHealth", 8.0, 0.0, 36.0).b(this.holeCheck);
    BooleanSetting crystalCalculate = this.bsetting("CalculateDmg", true);
    DoubleSetting maxSelfDmg = this.dsetting("MaxSelfDmg", 26.0, 0.0, 36.0).b(this.crystalCalculate);

    @Override
    public void onUpdate() {
        boolean shouldSwitch;
        Item sOffhandItem;
        if (SmartOffHand.fullNullCheck()) {
            return;
        }
        if (!AutoTotem.INSTANCE.soft.getValue().booleanValue()) {
            AutoTotem.INSTANCE.soft.setValue(true);
        }
        int crystals = SmartOffHand.mc.player.inventory.mainInventory.stream().filter(itemStack -> itemStack.getItem() == Items.END_CRYSTAL).mapToInt(ItemStack::getCount).sum();
        if (SmartOffHand.mc.player.getHeldItemOffhand().getItem() == Items.END_CRYSTAL) {
            crystals += SmartOffHand.mc.player.getHeldItemOffhand().getCount();
        }
        int gapple = SmartOffHand.mc.player.inventory.mainInventory.stream().filter(itemStack -> itemStack.getItem() == Items.GOLDEN_APPLE).mapToInt(ItemStack::getCount).sum();
        if (SmartOffHand.mc.player.getHeldItemOffhand().getItem() == Items.GOLDEN_APPLE) {
            gapple += SmartOffHand.mc.player.getHeldItemOffhand().getCount();
        }
        this.totems = SmartOffHand.mc.player.inventory.mainInventory.stream().filter(itemStack -> itemStack.getItem() == Items.TOTEM_OF_UNDYING).mapToInt(ItemStack::getCount).sum();
        if (SmartOffHand.mc.player.getHeldItemOffhand().getItem() == Items.TOTEM_OF_UNDYING) {
            ++this.totems;
        }
        Item item = null;
        if (!SmartOffHand.mc.player.getHeldItemOffhand().isEmpty()) {
            item = SmartOffHand.mc.player.getHeldItemOffhand().getItem();
        }
        this.count = item != null ? (item.equals(Items.END_CRYSTAL) ? crystals : (item.equals(Items.TOTEM_OF_UNDYING) ? this.totems : gapple)) : 0;
        Item handItem = SmartOffHand.mc.player.getHeldItemMainhand().getItem();
        Item offhandItem = ((Mode)((Object)this.mode.getValue())).equals((Object)Mode.Crystal) ? Items.END_CRYSTAL : Items.GOLDEN_APPLE;
        Item item2 = sOffhandItem = ((Mode)((Object)this.mode.getValue())).equals((Object)Mode.Crystal) ? Items.GOLDEN_APPLE : Items.END_CRYSTAL;
        if (((SMode)((Object)this.switchMode.getValue())).equals((Object)SMode.Sword)) {
            shouldSwitch = SmartOffHand.mc.player.getHeldItemMainhand().getItem() instanceof ItemSword && Mouse.isButtonDown((int)1) && (Boolean)this.autoSwitch.getValue() != false;
        } else {
            boolean bl = shouldSwitch = Mouse.isButtonDown((int)1) && (Boolean)this.autoSwitch.getValue() != false && !(handItem instanceof ItemFood) && !(handItem instanceof ItemExpBottle) && !(handItem instanceof ItemBlock);
        }
        if (this.shouldTotem() && this.getItemSlot(Items.TOTEM_OF_UNDYING) != -1) {
            this.switch_Totem();
        } else if (shouldSwitch && this.getItemSlot(sOffhandItem) != -1) {
            if (!SmartOffHand.mc.player.getHeldItemOffhand().getItem().equals(sOffhandItem)) {
                int slot = this.getItemSlot(sOffhandItem) < 9 ? this.getItemSlot(sOffhandItem) + 36 : this.getItemSlot(sOffhandItem);
                this.switchTo(slot);
            }
        } else if (this.getItemSlot(offhandItem) != -1) {
            int slot;
            int n = slot = this.getItemSlot(offhandItem) < 9 ? this.getItemSlot(offhandItem) + 36 : this.getItemSlot(offhandItem);
            if (!SmartOffHand.mc.player.getHeldItemOffhand().getItem().equals(offhandItem)) {
                this.switchTo(slot);
            }
        } else {
            this.switch_Totem();
        }
    }

    public boolean shouldTotem() {
        if (((Boolean)this.totem.getValue()).booleanValue()) {
            return this.checkHealth() || SmartOffHand.mc.player.getItemStackFromSlot(EntityEquipmentSlot.CHEST).getItem() == Items.ELYTRA && (Boolean)this.elytra.getValue() != false || SmartOffHand.mc.player.fallDistance >= 5.0f || HoleUtil.isPlayerInHole() && (Boolean)this.holeCheck.getValue() != false && (double)(SmartOffHand.mc.player.getHealth() + SmartOffHand.mc.player.getAbsorptionAmount()) <= (Double)this.holeSwitch.getValue() || (Boolean)this.crystalCalculate.getValue() != false && this.calcHealth();
        }
        return false;
    }

    public boolean calcHealth() {
        double maxDmg = 0.5;
        for (final Entity entity : new ArrayList<Entity>(SmartOffHand.mc.world.loadedEntityList)) {
            if (!(entity instanceof EntityEnderCrystal)) {
                continue;
            }
            if (SmartOffHand.mc.player.getDistance(entity) > 12.0f) {
                continue;
            }
            final double d = CrystalUtil.calculateDamage(entity.posX, entity.posY, entity.posZ, (Entity)SmartOffHand.mc.player);
            if (d <= maxDmg) {
                continue;
            }
            maxDmg = d;
        }
        return maxDmg - 0.5 > SmartOffHand.mc.player.getHealth() + SmartOffHand.mc.player.getAbsorptionAmount() || maxDmg > this.maxSelfDmg.getValue();
    }

    public boolean checkHealth() {
        boolean lowHealth = (double)(SmartOffHand.mc.player.getHealth() + SmartOffHand.mc.player.getAbsorptionAmount()) <= (Double)this.sbHealth.getValue();
        boolean notInHoleAndLowHealth = lowHealth && !HoleUtil.isPlayerInHole();
        return (Boolean)this.holeCheck.getValue() != false ? notInHoleAndLowHealth : lowHealth;
    }

    public void switch_Totem() {
        if (this.totems != 0 && !SmartOffHand.mc.player.getHeldItemOffhand().getItem().equals(Items.TOTEM_OF_UNDYING)) {
            int slot = this.getItemSlot(Items.TOTEM_OF_UNDYING) < 9 ? this.getItemSlot(Items.TOTEM_OF_UNDYING) + 36 : this.getItemSlot(Items.TOTEM_OF_UNDYING);
            this.switchTo(slot);
        }
    }

    public void switchTo(int slot) {
        try {
            if (timerUtils.passed(this.delay.getValue())) {
                SmartOffHand.mc.playerController.windowClick(SmartOffHand.mc.player.inventoryContainer.windowId, slot, 0, ClickType.PICKUP, (EntityPlayer)SmartOffHand.mc.player);
                SmartOffHand.mc.playerController.windowClick(SmartOffHand.mc.player.inventoryContainer.windowId, 45, 0, ClickType.PICKUP, (EntityPlayer)SmartOffHand.mc.player);
                SmartOffHand.mc.playerController.windowClick(SmartOffHand.mc.player.inventoryContainer.windowId, slot, 0, ClickType.PICKUP, (EntityPlayer)SmartOffHand.mc.player);
                timerUtils.reset();
            }
        }
        catch (Exception exception) {
            // empty catch block
        }
    }

    public int getItemSlot(Item input) {
        int itemSlot = -1;
        for (int i = 45; i > 0; --i) {
            if (SmartOffHand.mc.player.inventory.getStackInSlot(i).getItem() != input) continue;
            itemSlot = i;
            break;
        }
        return itemSlot;
    }

    @Override
    public String getHudInfo() {
        if (SmartOffHand.mc.player.getHeldItemOffhand().getItem() == Items.TOTEM_OF_UNDYING) {
            return "Totem";
        }
        if (SmartOffHand.mc.player.getHeldItemOffhand().getItem() == Items.END_CRYSTAL) {
            return "Crystal";
        }
        if (SmartOffHand.mc.player.getHeldItemOffhand().getItem() == Items.GOLDEN_APPLE) {
            return "Gapple";
        }
        if (SmartOffHand.mc.player.getHeldItemOffhand().getItem() == Items.BED) {
            return "Bed";
        }
        return "None";
    }

    public static enum SMode {
        RClick,
        Sword;

    }

    public static enum Mode {
        Crystal,
        Gap;

    }
}

File: src/main/java/me/windyteam/kura/module/modules/xddd/SmartOffHand.kt
package me.windyteam.kura.module.modules.xddd

import me.windyteam.kura.module.Category
import me.windyteam.kura.module.Module
import me.windyteam.kura.module.modules.combat.AutoTotem
import me.windyteam.kura.utils.TimerUtils
import me.windyteam.kura.utils.entity.CrystalUtil
import me.windyteam.kura.utils.entity.HoleUtil
import net.minecraft.entity.Entity
import net.minecraft.entity.item.EntityEnderCrystal
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.init.Items
import net.minecraft.inventory.ClickType
import net.minecraft.inventory.EntityEquipmentSlot
import net.minecraft.item.*
import org.lwjgl.input.Mouse

@Module.Info(name = "SmartOffHand", category = Category.XDDD, description = "StupidOffHand")
class SmartOffHand : Module() {
    private var totems = 0
    private var count = 0
    private var mode = msetting("Mode", Mode.Crystal)
    private var delay = isetting("Delay", 0, 0, 1000)
    private var totem = bsetting("SwitchTotem", true)
    private var sbHealth = dsetting("Health", 11.0, 0.0, 36.0)
    private var autoSwitch = bsetting("SwitchGap", true)
    private var switchMode = msetting("GapWhen", SMode.RClick).b(autoSwitch)
    private var elytra = bsetting("CheckElytra", true)
    private var holeCheck = bsetting("CheckHole", false)
    private var holeSwitch = dsetting("HoleHealth", 8.0, 0.0, 36.0).b(holeCheck)
    private var crystalCalculate = bsetting("CalculateDmg", true)
    private var maxSelfDmg = dsetting("MaxSelfDmg", 26.0, 0.0, 36.0).b(crystalCalculate)
    override fun onUpdate() {
        val shouldSwitch: Boolean
        val sOffhandItem: Item
        if (fullNullCheck()) {
            return
        }
        if (!AutoTotem.INSTANCE.soft.value) {
            AutoTotem.INSTANCE.soft.value = true
        }
        var crystals = mc.player.inventory.mainInventory.stream()
            .filter { itemStack: ItemStack -> itemStack.getItem() === Items.END_CRYSTAL }
            .mapToInt { obj: ItemStack -> obj.count }.sum()
        if (mc.player.heldItemOffhand.getItem() === Items.END_CRYSTAL) {
            crystals += mc.player.heldItemOffhand.count
        }
        var gapple = mc.player.inventory.mainInventory.stream()
            .filter { itemStack: ItemStack -> itemStack.getItem() === Items.GOLDEN_APPLE }
            .mapToInt { obj: ItemStack -> obj.count }.sum()
        if (mc.player.heldItemOffhand.getItem() === Items.GOLDEN_APPLE) {
            gapple += mc.player.heldItemOffhand.count
        }
        totems = mc.player.inventory.mainInventory.stream()
            .filter { itemStack: ItemStack -> itemStack.getItem() === Items.TOTEM_OF_UNDYING }
            .mapToInt { obj: ItemStack -> obj.count }.sum()
        if (mc.player.heldItemOffhand.getItem() === Items.TOTEM_OF_UNDYING) {
            ++totems
        }
        var item: Item? = null
        if (!mc.player.heldItemOffhand.isEmpty()) {
            item = mc.player.heldItemOffhand.getItem()
        }
        count =
            if (item != null) (if (item == Items.END_CRYSTAL) crystals else if (item == Items.TOTEM_OF_UNDYING) totems else gapple) else 0
        val handItem = mc.player.heldItemMainhand.getItem()
        val offhandItem = if (mode.value as Mode? == Mode.Crystal as Any) Items.END_CRYSTAL else Items.GOLDEN_APPLE
        sOffhandItem = if (mode.value as Mode? == Mode.Crystal as Any) Items.GOLDEN_APPLE else Items.END_CRYSTAL
        val item2 = sOffhandItem
        if (switchMode.value as SMode? == SMode.Sword as Any) {
            shouldSwitch = mc.player.heldItemMainhand.getItem() is ItemSword && Mouse.isButtonDown(
                1
            ) && autoSwitch.value as Boolean != false
        } else {
            shouldSwitch =
                Mouse.isButtonDown(1) && autoSwitch.value as Boolean != false && handItem !is ItemFood && handItem !is ItemExpBottle && handItem !is ItemBlock
            val bl = shouldSwitch
        }
        if (shouldTotem() && getItemSlot(Items.TOTEM_OF_UNDYING) != -1) {
            switch_Totem()
        } else if (shouldSwitch && getItemSlot(sOffhandItem) != -1) {
            if (mc.player.heldItemOffhand.getItem() != sOffhandItem) {
                val slot =
                    if (getItemSlot(sOffhandItem) < 9) getItemSlot(sOffhandItem) + 36 else getItemSlot(sOffhandItem)
                switchTo(slot)
            }
        } else if (getItemSlot(offhandItem) != -1) {
            val slot: Int
            slot = if (getItemSlot(offhandItem) < 9) getItemSlot(offhandItem) + 36 else getItemSlot(offhandItem)
            val n = slot
            if (mc.player.heldItemOffhand.getItem() != offhandItem) {
                switchTo(slot)
            }
        } else {
            switch_Totem()
        }
    }

    private fun shouldTotem(): Boolean {
        return if (totem.value as Boolean) {
            checkHealth() || mc.player.getItemStackFromSlot(EntityEquipmentSlot.CHEST)
                .getItem() === Items.ELYTRA && elytra.value as Boolean || mc.player.fallDistance >= 5.0f || HoleUtil.isPlayerInHole() && holeCheck.value as Boolean && (mc.player.health + mc.player.getAbsorptionAmount()).toDouble() <= holeSwitch.value as Double || crystalCalculate.value as Boolean && calcHealth()
        } else false
    }

    fun calcHealth(): Boolean {
        var maxDmg = 0.5
        for (entity in ArrayList(mc.world.loadedEntityList)) {
            if (entity !is EntityEnderCrystal) {
                continue
            }
            if (mc.player.getDistance(entity) > 12.0f) {
                continue
            }
            val d = CrystalUtil.calculateDamage(entity.posX, entity.posY, entity.posZ, mc.player as Entity).toDouble()
            if (d <= maxDmg) {
                continue
            }
            maxDmg = d
        }
        return maxDmg - 0.5 > mc.player.health + mc.player.getAbsorptionAmount() || maxDmg > maxSelfDmg.value
    }

    fun checkHealth(): Boolean {
        val lowHealth = (mc.player.health + mc.player.getAbsorptionAmount()).toDouble() <= sbHealth.value as Double
        val notInHoleAndLowHealth = lowHealth && !HoleUtil.isPlayerInHole()
        return if (holeCheck.value as Boolean != false) notInHoleAndLowHealth else lowHealth
    }

    fun switch_Totem() {
        if (totems != 0 && mc.player.heldItemOffhand.getItem() != Items.TOTEM_OF_UNDYING) {
            val slot =
                if (getItemSlot(Items.TOTEM_OF_UNDYING) < 9) getItemSlot(Items.TOTEM_OF_UNDYING) + 36 else getItemSlot(
                    Items.TOTEM_OF_UNDYING
                )
            switchTo(slot)
        }
    }

    fun switchTo(slot: Int) {
        try {
            if (timerUtils.passed(delay.value)) {
                mc.playerController.windowClick(
                    mc.player.inventoryContainer.windowId,
                    slot,
                    0,
                    ClickType.PICKUP,
                    mc.player as EntityPlayer
                )
                mc.playerController.windowClick(
                    mc.player.inventoryContainer.windowId,
                    45,
                    0,
                    ClickType.PICKUP,
                    mc.player as EntityPlayer
                )
                mc.playerController.windowClick(
                    mc.player.inventoryContainer.windowId,
                    slot,
                    0,
                    ClickType.PICKUP,
                    mc.player as EntityPlayer
                )
                timerUtils.reset()
            }
        } catch (exception: Exception) {
            // empty catch block
        }
    }

    fun getItemSlot(input: Item): Int {
        var itemSlot = -1
        for (i in 45 downTo 1) {
            if (mc.player.inventory.getStackInSlot(i).getItem() !== input) continue
            itemSlot = i
            break
        }
        return itemSlot
    }

    override fun getHudInfo(): String {
        if (mc.player.heldItemOffhand.getItem() === Items.TOTEM_OF_UNDYING) {
            return "Totem"
        }
        if (mc.player.heldItemOffhand.getItem() === Items.END_CRYSTAL) {
            return "Crystal"
        }
        if (mc.player.heldItemOffhand.getItem() === Items.GOLDEN_APPLE) {
            return "Gapple"
        }
        return if (mc.player.heldItemOffhand.getItem() === Items.BED) {
            "Bed"
        } else "None"
    }

    enum class SMode {
        RClick, Sword
    }

    enum class Mode {
        Crystal, Gap
    }

    companion object {
        var timerUtils = TimerUtils()
    }
}